ADR/0039-creating-pending-pipelineruns.md

```diff
@@ -0,0 +1,66 @@
+# 39. Creating pending pipeline runs
+
+Date: 2024-11-21
+
+## Status
+
+Accepted
+
+## Context
+
+There is a frequent use case amongst Konflux users when triggering large amount of builds: sometimes users do not care how quickly the pipeline run will finish, but they want it to finish eventually.
+Now what happens is that part of the pipeline runs started in this big bulk is blocked by unavailable resources (e.g. namespace memory limit quotas or MPC pool size).
+Because of how Tekton works, no matter what these pipeline runs are considered running (even if their pods are waiting for resources) and so eventually timeout setting strikes.
+
+So say you started 300 pipeline runs at Fry evening and you expect to see 300 new images at Mon morning.
+You have resource quota for 20 concurrent builds, so 20 builds get started, when they finish 20 more and so on.
+But because all 300 pipeline runs started (from Tekton point of view) when they were created, after few iterations all the remaining pipeline runs will be cancelled (as timeouted).
+
+Another use case for what we are proposing is protecting cluster (e.g. etcd) from spikes of running lots of concurrent pipeline runs.
+
+## Decision
+
+We will use idea of creating pipeline runs in pending state and having a standalone controller that would implement any custom logic on when to start each pending pipeline:
+
+1. We add `.spec.status: "PipelineRunPending"` to pipeline run YAMLs.
+1. We will implement controller, that will contain code that will decide when to start (by just nulling `.spec.status` field - then Pipelines Controller takes over) what pipeline run. Some possibilities come to mind:
+   1. Controller will keep an eye on current number of concurrently running pipeline runs in the cluster (using Prometheus metric `avg(tekton_pipelines_controller_running_pipelineruns)`), and if that is below given threshold, it starts the pipeline run(s).
+   1. If pipeline run have `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` annotation, controller checks if MPC have free capacity (using max for the pool and Prometheus metric `multi_platform_controller_running_tasks`) for mentioned platforms and if so, will start it.
```

Comment:
Is this an annotation that is used today or is it a new proposed one? Can it be generalized as the multi-platform-controller isn't a core component?

The specific annotation values are implementations of individual Konflux deployments.

Reply:
It is just a random string I used in tests in Stage we did.

I was thinking about this as sort of plugable system:

* `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` - means you want your pipeline run to be gated based on MPC pool availability
* `prsstarter.example.com/memory_limit_quota: 10GiB` - means you want your pipeline to be started only if your namespace have at least 10GiB free in memory limit quota
* `prsstarter.example.com/other_plugin: other_plugin_configuration` - means you want your pipeline to be gated by some other plugin

Should I just avoid specific annotations or stress it is just an example? How can I improve it in the doc?